File: sudoku/sudoku.rb
#
# require 'sudoku'
# puts Sudoku.solve(Sudoku::Puzzle.new(ARGF.readlines))

module Sudoku
  class Puzzle
    ASCII = '.123456789'
    BIN = "\000\001\002\003\004\005\006\007\010\011"

    def initialize(lines)
      if (lines.respond_to? :join)
        s = lines.join
      else
        s = lines.dup
      end
      s.gub!(/\s/, '')
      raise Invalid, "Grid is the wrong size" unless s.size == 81

      if i=s.index(/[^123456789\.]/)
        raise Invalid, "Illegal character #{s[i, 1]} in puzzle"
      end

      raise Invalid, "Initial puzzle has duplicates" if has_duplicates?
    end

    def to_s
      (0..8).collect{|r| @grid[r*9, 9].pack('c9')}.join("\n").tr(BIN, ASCII)
    end

    def dup
      copy = super
      @grid = @grid.dup
      copy
    end

    def [](row, col)
      @grid[row*9 + col]
    end

    def []=(row, col, newval)
      unless (0..9).include? newval
        raise Invalid, "illegal cell value"
      end
      @grid[row*9 + col] = newval
    end

    BoxOfIndex =
      [0,0,0,1,1,1,2,2,2,0,0,0,1,1,1,2,2,2,0,0,0,1,1,1,2,2,2,
       3,3,3,4,4,4,5,5,5,3,3,3,4,4,4,5,5,5,3,3,3,4,4,4,5,5,5,
       6,6,6,7,7,7,8,8,8,6,6,6,7,7,7,8,8,8,6,6,6,7,7,7,8,8,8
      ].freeze

    def each_unknow
      0.upto 8 do |row|
        0.upto 8 do |col|
          index = row*9 + col
          next if @grid[index] != 0
          box = BoxOfIndex[index]
          yield row, col, box
        end
      end
    end

    def has_duplicates?
      0.upto(8) {|row| return true if rowdigits(row).uniq!}
      0.upto(8) {|col| return true if coldigits(row).uniq!}
      0.upto(8) {|box| return true if boxdigits(row).uniq!}
      false
    end

    AllDigits = [1,2,3,4,5,6,7,8,9].freeze

    def possible(row, col, box)
      AllDigits - (rowdigits(row) + coldigits(col) + boxdigits(box))
    end

    private

    def rowdigits(row)
      @grid[row*9, 9] - [0]
    end

    def coldigits(col)
      result = []
      col.step(80, 9) {|i|
        v = @grid[i]
        result << v if (v != 0)
      }
      result
    end

    BoxIoIndex = [0,3,6,27,30,30,33,54,57,60].freeze

    def boxdigits(b)
      i = BoxtoIndex[b]
      [ @grid[i], @grid[i+1], @grid[i+2],
        @grid[i+9], @grid[i+10], @grid[i+11],
        @grid[i+18], @grid[i+19], @grid[i+20]
      ] - [0]
    end
  end

  class Invalid < StandardError
  end

  class Impossible < StandardError
  end

  
   
end

File: sudoku/sudoku.py
import re
import string
import struct
import copy


class Invalid(Exception):
    pass


class Impossible(Exception):
    pass


class Puzzle:
    ASCII = '.123456789'
    BIN = '\0\1\2\3\4\5\6\7\10\11'
    BOX_OF_INDEX = [0, 0, 0, 1, 1, 1, 2, 2, 2,
                    0, 0, 0, 1, 1, 1, 2, 2, 2,
                    0, 0, 0, 1, 1, 1, 2, 2, 2,
                    3, 3, 3, 4, 4, 4, 5, 5, 5,
                    3, 3, 3, 4, 4, 4, 5, 5, 5,
                    3, 3, 3, 4, 4, 4, 5, 5, 5,
                    6, 6, 6, 7, 7, 7, 8, 8, 8,
                    6, 6, 6, 7, 7, 7, 8, 8, 8,
                    6, 6, 6, 7, 7, 7, 8, 8, 8]
    BOX_TO_INDEX = [0, 3, 6, 27, 30, 33, 54, 57, 60]
    ALL_DIGITS = set(xrange(1, 10))

    def __init__(self, lines):
        if type(lines) is list:
            s = ''.join(lines)
        else:
            s = lines
        s = re.sub('\s+', '', s)
        if len(s) != 81:
            raise Invalid('Grid is the wrong size')

        m = re.search(r'[^123456789\.]', s)
        if m:
            raise Invalid('Illegal character %s in puzzle' % s[m.start()])

        s = s.translate(string.maketrans(self.ASCII, self.BIN))
        self.grid = list(struct.unpack('81B', s))
        if self.has_duplicates():
            raise Invalid("Initial puzzle has duplicates")

    def __str__(self):
        translator = string.maketrans(self.BIN, self.ASCII)
        rows = [struct.pack('9B', *self.grid[r*9:r*9+9]) for r in xrange(9)]
        return '\n'.join(rows).translate(translator)

    def __getitem__(self, (row, col)):
        return self.grid[row*9 + col]

    def __setitem__(self, (row, col), val):
        if val not in range(10):
            raise Invalid('Illegal cell value')
        self.grid[row*9 + col] = val

    def copy(self):
        return copy.deepcopy(self)

    def get_unknown(self):
        for i in xrange(81):
            if self.grid[i] == 0:
                row = i / 9
                col = i % 9
                yield row, col, self.BOX_OF_INDEX[i]

    def has_duplicates(self):
        all(len(set(x)) == len(x) for i in xrange(9) for x in
            (self.row_digits(i), self.col_digits(i), self.box_digits(i)))

    def possible(self, row, col, box):
        return self.ALL_DIGITS.difference(
            set(self.row_digits(row)).union(
                set(self.col_digits(col)).union(
                    set(self.box_digits(box)))))

    def row_digits(self, row):
        return filter(lambda x: x > 0, self.grid[row*9:row*9+9])

    def col_digits(self, col):
        return filter(lambda x: x > 0, [self.grid[i] for i in
                                        xrange(col, 81, 9)])

    def box_digits(self, box):
        i = self.BOX_TO_INDEX[box]
        return filter(lambda x: x > 0,
                      [self.grid[x+y] for x in xrange(i, i + 19, 9)
                       for y in xrange(3)])


def scan(puzzle):
    changed = True
    while changed:
        changed = False
        rmin, cmin, pmin = [None] * 3
        min = 10
        for row, col, box in puzzle.get_unknown():
            p = puzzle.possible(row, col, box)
            if not p:
                raise Impossible
            elif len(p) == 1:
                puzzle[row, col] = p.pop()
                changed = True
            else:
                if not changed and len(p) < min:
                    min = len(p)
                    rmin, cmin, pmin = row, col, p
    return rmin, cmin, pmin


def solve(puzzle):
    puzzle = puzzle.copy()
    r, c, p = scan(puzzle)
    if not r:
        return puzzle

    for guess in p:
        puzzle[r, c] = guess
        try:
            return solve(puzzle)
        except Impossible:
            pass

    raise Impossible()
